crates/telio-firewall/src/firewall.rs

```diff
@@ -938,6 +986,202 @@ impl StatefullFirewall {
+
+                    tcppkg.set_checksum(pnet_packet::tcp::ipv4_checksum(
+                        &tcppkg.to_immutable(),
+                        &src,
+                        &dst,
+                    ));
+
+                    ipv4pkg.set_checksum(0);
+                    ipv4pkg.set_checksum(pnet_packet::ipv4::checksum(&ipv4pkg.to_immutable()));
+
+                    telio_log_debug!("Injecting IPv4 TCP RST packet {key:#?}");
+
+                    #[allow(index_access_check)]
+                    ipv4pkgbuf[..IPV4_LEN].copy_from_slice(ipv4pkg.packet());
+
+                    #[allow(index_access_check)]
+                    ipv4pkgbuf[IPV4_LEN..].copy_from_slice(tcppkg.packet());
+
+                    sink4.write_all(&ipv4pkgbuf)?;
+                }
+                (IpAddr::Ipv6(_), IpAddr::Ipv6(_)) => (), // TODO(msz): implement this pice when IPv4 will be fully supported
+                _ => telio_log_warn!(
+                    "Local and remote IP addrs version missmatch, this should never happen"
+                ),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn reset_udp_conns(
+        &self,
+        pubkey: &PublicKey,
+        endpoint_ip: Option<StdIpAddr>,
+        sink4: &mut dyn io::Write,
+    ) -> io::Result<()> {
+        let Ok(mut udp_conn_cache) = self.udp.lock() else {
+            telio_log_error!("UDP cache poisoned");
+            return Ok(());
+        };
+
+        telio_log_debug!(
+            "Inspecting {} UDP connections for reset",
+            udp_conn_cache.len()
+        );
+
+        const IPV4_HEADER_LEN: usize = 20;
+        const ICMP_HEADER_LEN: usize = 8;
+        const ICMP_DATA_LEN: usize = UdpConnectionInfo::LAST_PKG_HEADER;
+
+        let mut ipv4pkgbuf = [0u8; IPV4_HEADER_LEN + ICMP_HEADER_LEN + ICMP_DATA_LEN];
+
+        // Write most of the fields for ICMP and IP packets upfront
+        #[allow(clippy::expect_used)]
+        #[allow(index_access_check)]
+        let mut ipv4pkg = MutableIpv4Packet::new(&mut ipv4pkgbuf[..IPV4_HEADER_LEN])
+            .expect("IPv4 buffer should not be too small");
+        ipv4pkg.set_version(4);
+        ipv4pkg.set_header_length(5);
+        ipv4pkg.set_total_length((IPV4_HEADER_LEN + ICMP_HEADER_LEN + ICMP_DATA_LEN) as _);
+        ipv4pkg.set_flags(Ipv4Flags::DontFragment);
+        ipv4pkg.set_ttl(0xFF);
+        ipv4pkg.set_next_level_protocol(IpNextHeaderProtocols::Icmp);
+        drop(ipv4pkg);
+
+        #[allow(clippy::expect_used)]
+        #[allow(index_access_check)]
+        let mut icmppkg = MutableIcmpPacket::new(
+            &mut ipv4pkgbuf[IPV4_HEADER_LEN..(IPV4_HEADER_LEN + ICMP_HEADER_LEN)],
+        )
+        .expect("ICMP buffer should not be too small");
+        icmppkg.set_icmp_type(IcmpTypes::DestinationUnreachable);
+        icmppkg.set_icmp_code(IcmpCodes::DestinationPortUnreachable);
+        drop(icmppkg);
+
+        let iter = udp_conn_cache.iter().filter_map(|(k, v)| {
```

Comment:
If we have that `continue` on line 1140, maybe this filtering is not really necessary? (also applies to TCP).